gateway/api/views.py

```diff
@@ -114,16 +116,63 @@ def get_model_serializer_run_program(*args, **kwargs):
     def get_serializer_class(self):
         return self.serializer_class
 
+    def get_object(self):
+        logger.warning("ProgramViewSet.get_object not implemented")
+
     def get_queryset(self):
-        # Allow unauthenticated users to read the swagger documentation
-        if self.request.user is None or not self.request.user.is_authenticated:
-            return Program.objects.none()
-        return (
-            Program.objects.all().filter(author=self.request.user).order_by("-created")
+        author = self.request.user
+
+        logger.info("ProgramViewSet get view_program permission")
+        view_program_permission = Permission.objects.get(
+            codename=VIEW_PROGRAM_PERMISSION
         )
 
-    def perform_create(self, serializer):
-        serializer.save(author=self.request.user)
+        # Groups logic
+        user_criteria = Q(user=author)
+        author_criteria = Q(author=author)
+        view_permission_criteria = Q(permissions=view_program_permission)
+        author_groups_with_view_permissions = Group.objects.filter(
+            user_criteria & view_permission_criteria
+        )
+        author_groups_with_view_permissions_count = Group.objects.filter(
```

Comment:
looks like we can save one query by using
```python
author_groups_with_view_permissions_count = author_groups_with_view_permissions.count()
```

Reply:
Good point, I think you are right! Let me double check and change it!

Reply:
Done in [6ce8d69](https://github.com/Qiskit-Extensions/quantum-serverless/pull/1306/commits/6ce8d69d39d6d8b95f714585253f123173bdcb65)